### PR TITLE
Set Referrer in the browser process

### PR DESCRIPTION
**Pull request by bbondy**
Fix Fix brave/brave-browser#185

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions

### CODE REVIEW

File: browser/net/brave_site_hacks_network_delegate_helper.cc [hunk 1]
```diff
@@ -112,6 +126,36 @@ bool IsBlockTwitterSiteHack(net::URLRequest* request,
...
+  headers->GetHeader(kRefererHeader, &referrer);
+  if (headers->GetHeader(kRefererHeader, &referrer) &&
+      !IsWhitelistedReferrer(tab_origin, target_origin) &&
+      !allow_referrers &&
+      shields_up &&
+      !SameDomainOrHost(target_origin, GURL(referrer),
+          INCLUDE_PRIVATE_REGISTRIES)) {
+    auto referrer = Referrer::SanitizeForRequest(target_origin,
+        Referrer(target_origin,
+            request->referrer_policy() == net::URLRequest::NO_REFERRER ?
```
Comment (bridiver) [inline browser/net/brave_site_hacks_network_delegate_helper.cc]: I don't think this is the correct thing to do here. `SanitizeForRequest` should handle this if you just pass the `referrer_policy`

File: components/brave_shields/browser/brave_shields_web_contents_observer.cc [hunk 2]
```diff
@@ -250,4 +261,58 @@ void BraveShieldsWebContentsObserver::RegisterProfilePrefs(
...
+  }
+  GURL target_origin = navigation_handle->GetURL().GetOrigin();
+  if (!navigation_entry) {
+    return;
+  }
+  Profile* profile =
+      Profile::FromBrowserContext(web_contents()->GetBrowserContext());
+  GURL tab_origin(navigation_entry->GetURL().GetOrigin());
+
+  std::unique_ptr<base::Value> referrer_value =
```
Comment (bridiver) [inline components/brave_shields/browser/brave_shields_web_contents_observer.cc]: seems like we have some duplicated logic with `ApplyPotentialReferrerBlock` that could get out-of-sync. Maybe we should just return the updated (possibly unchanged) referrer so we can use it in both places?

File: components/brave_shields/browser/brave_shields_web_contents_observer.cc [hunk 3]
```diff
@@ -250,4 +261,58 @@ void BraveShieldsWebContentsObserver::RegisterProfilePrefs(
...
+      target_origin, INCLUDE_PRIVATE_REGISTRIES);
+
+  if (referrer_setting != CONTENT_SETTING_ALLOW &&
+      shields_setting != CONTENT_SETTING_BLOCK &&
+      !sameTLDP1 &&
+      !brave::IsWhitelistedReferrer(tab_origin, target_origin) &&
+      !original_referrer.url.is_empty()) {
+    auto referrer = original_referrer;
+    referrer.url = navigation_handle->GetURL().GetOrigin();
+    navigation_entry->SetReferrer(
```
Comment (bridiver) [inline components/brave_shields/browser/brave_shields_web_contents_observer.cc]: this is an example of something that is different between the network delegate and the observer. I think this one is correct

File: components/brave_shields/browser/brave_shields_web_contents_observer.cc (hunk 2 above)
Comment (bbondy) [reply components/brave_shields/browser/brave_shields_web_contents_observer.cc]: This was refactored, but comment is old.

**Pull request merged**